### ISSUE TITLE
Version 1.46 aws_route53_zone changed behavior which breaks a lot of configurations

### ISSUE THREAD

**Issue by voroniys** [terraform-providers/terraform-provider-aws#6535]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.10
+ provider.aws v1.46.0

### Affected Resource(s)

* aws_route53_zone
in previous versions (at least in 1.41) property name was returned without trailing dot, what allow to use it in creation certificates, dns records, etc.

### Terraform Configuration Files
For instance (related to issue #6533 )

resource "aws_acm_certificate" "my_cert" {
  domain_name       = "${var.dns_name}.${aws_route53_zone.myzone.name}"
  validation_method = "DNS"
}

This construction is failing now due to traling dot in ${aws_route53_zone.myzone.name}

### Expected Behavior

To maintain compatibility with previous versions and remove traling from the name property.
It is much easier to add a dot where it is needed, than remove it in all places where it is harmful.

**Comment by brianabston001** [comment]
Interesting.. I have mine pinned to version 1.40.0 and I am having the issue.

**Comment by voroniys** [comment]
A possible solution for the issue could be a creation of interpolation function to strip trailing dot.

**Comment by brianabston001** [comment]
To get around this I am having to do a replace on code which is not ideal.

resource "aws_acm_certificate" "integration_cert" {
  domain_name = "api.${replace(aws_route53_zone.public_io.name, "/.$/", "")}"
  validation_method = "DNS"

  tags {
    Name        = "${var.environment}-integration-cert"
    Environment = "${var.environment}"
  }
}

**Comment by voroniys** [comment]
> Interesting.. I have mine pinned to version 1.40.0 and I am having the issue.

For me it is bigger issue, than certificates only. Between 1.41 and 1.42 the behaviour of property name has changed, so it is returning trailing dot now. I've pinned my version to 1.41 to avoid revriting of configurations that use this property. 
And there is no way to rewrite it BTW in many cases, as there is no function for sripping the trailing dot.

**Comment by voroniys** [comment]
> To get around this I am having to do a replace on code which is not ideal.
> 
> ```
>   domain_name = "api.${replace(aws_route53_zone.public_io.name, "/.$/", "")}"

I'm doing pretty much the same, but I'm using this name in dozens of the places, so I assign it once to local variable and than everywhere use ${local.dns_zone}.
And btw be careful - your regex is not correct - you removing any last symbol from the string. It should be /\\.$/ to remove last dot only.

**Comment by brianabston001** [comment]
> > To get around this I am having to do a replace on code which is not ideal.
> > ```
> >   domain_name = "api.${replace(aws_route53_zone.public_io.name, "/.$/", "")}"
> > ```
> 
> I'm doing pretty much the same, but I'm using this name in dozens of the places, so I assign it once to local variable and than everywhere use ${local.dns_zone}.
> And btw be careful - your regex is not correct - you removing any last symbol from the string. It should be /\.$/ to remove last dot only.

hmmm when I do an init with it like below I get an error `illegal char escape`

```
domain_name = "api.${replace(aws_route53_zone.public_io.name, "/\.$/", "")}"
```

Can you show me the example on how you are using it? I might just change over to that setup. I am using it almost a dozen times as well.

**Comment by voroniys** [comment]
locals {
  dns_zone = "${replace(aws_route53_zone.myzone.name, "/\\.$/", "")}"
}
resource "aws_acm_certificate" "3ds_cert" {
  domain_name       = "${var.dns_name}.${local.dns_zone}"
  subject_alternative_names = [
    "prefix1-${var.dns_name}.${local.dns_zone}",
    "prefix2-${var.dns_name}.${local.dns_zone}",
    "prefix3-mutual-${var.dns_name}.${local.dns_zone}"
  ]
  validation_method = "DNS"
  lifecycle {
    create_before_destroy = true
  }
}

**Comment by brianabston001** [comment]
> locals { dns_zone = "${replace(aws_route53_zone.myzone.name, "/\\.$/", "")}" } resource "aws_acm_certificate" "my_cert" { domain_name = "${var.dns_name}.${local.dns_zone}" subject_alternative_names = [ "prefix1-${var.dns_name}.${local.dns_zone}", "prefix2-${var.dns_name}.${local.dns_zone}", "prefix3-mutual-${var.dns_name}.${local.dns_zone}" ] validation_method = "DNS" lifecycle { create_before_destroy = true } }

Perfect.. That is cleaner than what I was doing... Also see what I was missing. Thanks!!

**Comment by horjulf** [comment]
The change was introduced [here](https://github.com/terraform-providers/terraform-provider-aws/commit/7351b747849ff8cf993b535ee49448314d776925#diff-ed533d79eb66d639b2497bb83c423eedL157).

@bflad was the behavior change intended ?

**Comment by bflad** [comment]
Thinking about this more, I should mention for the first option there, its not possible to fix the Terraform state to remove the trailing period based on the configuration if the state has already been refreshed to include the trailing period. The Terraform provider SDK does not give us a way to forcibly read (`d.Get()`, etc.) from only the Terraform configuration and not include the value from the Terraform state.

To put it another way, if you've already upgraded the provider and refreshed the Terraform state like #6533, the trailing period is "stuck" at this point even if you try to downgrade versions and refresh.

**Comment by horjulf** [comment]
I think a combination of both the 2nd and 3rd points is the way to go.
The unexpected change of behavior was the biggest impact IMO, having it mentioned in the Changelog and updating `aws_acm_certificate` to handle both cases should be enough to address this.

**Comment by horjulf** [comment]
Something to keep in mind is that the content of the field `name` is changed from its original value if created without the trailing dot (suppressRoute53ZoneNameWithTrailingDot), maybe a warning note should be added to the docs.

**Comment by xMordax** [comment]
Same here thing here for me, between 1.41 and 1.42 the behaviour of property name has changed and it adds a trailing dot in the end. Many of my certificates are broken right now.

**Comment by bflad** [comment]
Pull request submitted to have the `aws_acm_certificate` resource automatically trim a trailing period for `domain_name` and `subject_alternative_names`: #6844

**Comment by bflad** [comment]
The above has been released in [version 1.52.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1520-december-13-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by nicwise** [comment]
I had this on `aws_api_gateway_domain_name` too - the domain_name field doesn't like a `.` at the end. I'm guessing there might be more?
